plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Mappings.hs

```diff
@@ -0,0 +1,167 @@
+{- | Mappings.hs
+This module contains the mapping
+1. our token type from(to) LSP default token type
+2. from ghc type to our token type
+3. from hieAst identifier detail to our token type
+4. from lsp token to our token
+-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Ide.Plugin.SemanticTokens.Mappings where
+import qualified Data.List                       as List
+import           Data.List.Extra                 (chunksOf, (!?))
+import qualified Data.Map                        as Map
+import           Data.Maybe                      (fromMaybe)
+import qualified Data.Set                        as Set
+import           Data.Text                       (Text)
+import           Development.IDE.GHC.Compat
+import           Ide.Plugin.SemanticTokens.Types
+import           Language.LSP.Protocol.Types     (LspEnum (knownValues),
+                                                  SemanticTokenAbsolute (SemanticTokenAbsolute),
+                                                  SemanticTokenRelative (SemanticTokenRelative),
+                                                  SemanticTokenTypes (..),
+                                                  SemanticTokens (SemanticTokens),
+                                                  UInt, absolutizeTokens)
+
+{- |
+1. from our token type to LSP default token type
+-}
+
+semanticTokenAbsoluteActualToken :: SemanticTokenAbsolute -> ActualToken
+semanticTokenAbsoluteActualToken (SemanticTokenAbsolute line startChar len tokenType tokenModifiers) =
+    (line, startChar, len, fromLspTokenType tokenType, 0)
+
+-- mapping from our token type to LSP default token type
+toLspTokenType :: SemanticTokenType -> SemanticTokenTypes
+toLspTokenType tk = case tk of
+    -- TVariable     -> SemanticTokenTypes_Variable
+    -- left hand side of none pattern bind
+    TValBind      -> SemanticTokenTypes_Function
```

Comment:
I think I maybe need an example. Are we talking about something like this?
```
f a b c = ...
```
You want to highlight `f` differently from `a`/`b`/`c`?

I guess my argument is that this will (I think) highlight `let x = 1 in ...` with `x` as a "function", which it isn't. Hence why I'd be tempted to use "variable" for all of those. But using "parameter" for the function arguments seems nice.

Reply:
may be we can use variable for all val bindings
and leave pattern to be parameters :)

Reply:
also I a have an evil statement to make. 
given the fact that Haskell has no zero arity function. We have correspondence as:
f : f()
f x : f(x)
f x x : f(x,x)

Reply:
I think it would also be somewhat reasonable to use "function" for any binding, but I think that's arguably somewhat misleading... I'm also not _totally_ sold on using "parameter" for all pattern bindings. Many pattern bindings aren't function parameters! It might be trickier to find the bindings that really are function parameters.

Your screenshot is very helpful, thanks. It seems to me that why that the second picture is helpful is not so much that it distinguishes pattern bindings from other kinds of binding, but that it distinguishes _local_ names from _global_ names. It just happens that pattern bindings are always local and most local variables are pattern bindings. Unfortunately it looks like we don't have a good way to express this in the current standard set of token types. Perhaps we could suggest they add a `local` token modifier or something?

Apart from that I'm a bit torn - generally I think we should do what the spec says is correct and then if that shows up badly it's not our problem. On the other hand, sometimes it makes a big difference to do the "wrong" thing.

Reply:
I agree with you for the fact that we won't have a good way of expressing haskell token type to current standard set of token types. It is more of compromise.

Reply:
And also these mapping is temporary, 
I am digging more to seek, if there are a reasonable way of differing function, paramaters, and none function variable.

Reply:
Also it is more of a ghc problem for this. quoted of `Note [FunBind vs PatBind]` in ghc
```
{- Note [FunBind vs PatBind]
   ~~~~~~~~~~~~~~~~~~~~~~~~~
The distinction between FunBind and PatBind is a bit subtle. FunBind covers
patterns which resemble function bindings and simple variable bindings.

    f x = e
    f !x = e
    f = e
    !x = e          -- FunRhs has SrcStrict
    x `f` y = e     -- FunRhs has Infix

The actual patterns and RHSs of a FunBind are encoding in fun_matches.
The m_ctxt field of each Match in fun_matches will be FunRhs and carries
two bits of information about the match,

  * The mc_fixity field on each Match describes the fixity of the
    function binder in that match.  E.g. this is legal:
         f True False  = e1
         True `f` True = e2

  * The mc_strictness field is used /only/ for nullary FunBinds: ones
    with one Match, which has no pats. For these, it describes whether
    the match is decorated with a bang (e.g. `!x = e`).

By contrast, PatBind represents data constructor patterns, as well as a few
other interesting cases. Namely,

    Just x = e
    (x) = e
    x :: Ty = e
-}
```

Reply:
Also I am convinced that variable and function is more of a thing that user cares about than wether it is from value bindings or pattern bindings